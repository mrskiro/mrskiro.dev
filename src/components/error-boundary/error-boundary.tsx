import * as React from "react";
import * as Sentry from "@sentry/nextjs";

type Props = {
  children: React.ReactNode;
  fallback: (err: unknown) => React.ReactNode;
  onError?: (err: Error) => void;
};

type State = {
  cause: unknown | null;
};

export class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { cause: null };
  }
  static getDerivedStateFromError(cause: unknown) {
    return { cause };
  }
  componentDidCatch(err: unknown, _info: React.ErrorInfo) {
    if (err instanceof Error) {
      this.props.onError?.(err);
      Sentry.captureException(err);
    }
  }
  render() {
    if (this.state.cause) {
      return this.props.fallback(this.state.cause);
    }
    return this.props.children;
  }
}
