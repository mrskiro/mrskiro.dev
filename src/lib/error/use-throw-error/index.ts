export { useThrowError } from "./use-throw-error"
