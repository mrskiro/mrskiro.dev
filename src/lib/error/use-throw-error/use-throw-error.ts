import * as React from "react";

/**
 * for error boundary
 */
export const useThrowError = () => {
  const [err, setErr] = React.useState<Error | null>(null);
  const throwError = (err: Error) => {
    setErr(err);
  };
  if (err) {
    throw err;
  }
  return {
    throwError,
  };
};
