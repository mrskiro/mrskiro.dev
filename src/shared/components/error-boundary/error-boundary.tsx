import * as React from "react"

type Props = {
  children: React.ReactNode
  fallback: (err: unknown) => React.ReactNode
}

type State = {
  cause: unknown | null
}

export class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = { cause: null }
  }
  static getDerivedStateFromError(cause: unknown) {
    return { cause }
  }
  componentDidCatch(error: unknown, info: React.ErrorInfo) {
    console.log("from error boundary", { error, info })
  }
  render() {
    if (this.state.cause) {
      return this.props.fallback(this.state.cause)
    }
    return this.props.children
  }
}
