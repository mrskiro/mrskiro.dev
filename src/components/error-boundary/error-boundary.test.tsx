import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { useThrowError } from "@/lib/error/use-throw-error"
import { ErrorBoundary } from "./error-boundary"

it("ok", () => {
  const { container } = render(
    <ErrorBoundary fallback={() => <p>error</p>}>
      <h1>children</h1>
    </ErrorBoundary>
  )
  expect(screen.getByText("children")).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})

it("should display fallback when throw error in children", () => {
  jest.spyOn(console, "error").mockImplementation(() => {})
  const Component = () => {
    throw new Error("something error")
  }
  const { container } = render(
    <ErrorBoundary fallback={() => <p>error</p>}>
      <Component />
    </ErrorBoundary>
  )
  expect(screen.getByText("error")).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})

it("should display fallback when throw error in event handler", async () => {
  jest.spyOn(console, "error").mockImplementation(() => {})
  const user = userEvent.setup()
  const Component = () => {
    const { throwError } = useThrowError()
    return (
      <button
        type="button"
        onClick={() => {
          throwError(new Error("something event error"))
        }}
      >
        click me
      </button>
    )
  }
  render(
    <ErrorBoundary fallback={() => <p>error</p>}>
      <Component />
    </ErrorBoundary>
  )
  const btn = screen.getByRole("button")
  await user.click(btn)
  expect(screen.getByText("error")).toBeInTheDocument()
  expect(btn).not.toBeInTheDocument()
})

it("should call onError prop", () => {
  jest.spyOn(console, "error").mockImplementation(() => {})
  const onError = jest.fn()
  const ThrowError = () => {
    throw new Error("something error")
  }
  render(
    <ErrorBoundary fallback={() => <p>error</p>} onError={onError}>
      <ThrowError />
    </ErrorBoundary>
  )
  expect(onError).toHaveBeenCalled()
  expect(onError).toHaveBeenCalledWith(new Error("something error"))
})
