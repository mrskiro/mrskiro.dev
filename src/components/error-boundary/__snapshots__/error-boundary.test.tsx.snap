// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ok 1`] = `
<div>
  <h1>
    children
  </h1>
</div>
`;

exports[`should display fallback when throw error in children 1`] = `
<div>
  <p>
    error
  </p>
</div>
`;
